Fix authentication handlers using digest auth

sigmavirus24 opened the pull request:
Requests 2.8.0 introduced a thread-safe HTTPDigestAuth class which was
backwards incompatible for anything that attempts to work with it. This
adds a compatibility layer for convenience based on the version of
requests that is installed alongside the toolbelt.

Closes #108 

---

TODO:

- [x] Fix last test
- [ ] Add more tests around this
- [x] ???

sigmavirus24 commented:
[This](https://travis-ci.org/sigmavirus24/requests-toolbelt/jobs/87949403#L125) is exactly what I'm seeing locally. It should get set from [this conditional](https://github.com/sigmavirus24/requests-toolbelt/blob/master/requests_toolbelt/auth/http_proxy_digest.py#L68) but appears it isn't. I'm not certain why. I think this is @Lukasa's code so memories/context would be helpful.

Lukasa commented:
Yup, that's what I saw. My suspicion is that we're seeing some weird threading interaction but I'm not sure what yet.

sigmavirus24 commented:
@Lukasa so in the Proxy Digest Auth class, I had added the `init_per_thread_state()` in the `__call__` method (like it is in requests) and when I moved that to the `__init__` method, the test passed. This looks to have fixed it.

Lukasa commented:
Do you want to *also* add it to `__call__`? Because I think the issue here is that the auth object might be called in multiple threads *after* instantiation.

sigmavirus24 commented:
> Do you want to also add it to `__call__`? Because I think the issue here is that the auth object might be called in multiple threads after instantiation.

So that is why I thought it belonged in `__call__`. I'll see if the tests still pass with it in both locations (but to be clear the HTTPDigestAuth class only calls that in `__call__` so I'm baffled why we would need to call it any differently.

sigmavirus24 commented:
Eh, it works in both places and I realized that `stale_rejects` would make the proxy digest auth class not thread-safe so I took the liberty of adding extra code to protect against that by using thread local storage if it is available.

Care to give this another look over @Lukasa?

sigmavirus24 commented:
Also, I'm not sure what tests to write for this pull request.
